src/pflib/algorithm/toa_vref_scan.cxx

```diff
@@ -0,0 +1,91 @@
+#include "pflib/algorithm/toa_vref_scan.h"
+
+#include "pflib/utility/string_format.h"
+#include "pflib/utility/efficiency.h"
+
+#include "pflib/DecodeAndBuffer.h"
+
+/**
+ * calculate the highest TOA_VREF value for each link, for which there is a non-zero TOA efficiency
+ */
+static std::array<double, 72> get_toa_efficiencies(const std::vector<pflib::packing::SingleROCEventPacket> &data) {
+  std::array<double, 72> efficiencies;
+  /// reserve a vector of the appropriate size to avoid repeating allocation time for all 72 channels
+  std::vector<int> toas(data.size());
+  for (int ch{0}; ch < 72; ch++) {
+    for (std::size_t i{0}; i < toas.size(); i++) {
+      toas[i] = data[i].channel(ch).toa();
+    }
+    /// we assume that the data provided is not empty otherwise the efficiency calculation is meaningless
+    efficiencies[ch] = pflib::utility::efficiency(toas);
+  }
+  return efficiencies;
+}
+
+namespace pflib::algorithm {
+
+std::map<std::string, std::map<std::string, int>>
+toa_vref_scan(Target* tgt, ROC roc) {
+  static auto the_log_{::pflib::logging::get("toa_vref_scan")};
+
+  /// do a run of 100 samples per toa_vref to measure the TOA
+  /// efficiency when looking at pedestal data
+
+  static const std::size_t n_events = 100;
+
+  tgt->setup_run(1, Target::DaqFormat::SIMPLEROC, 1);
+
+  std::array<int, 2> target; //toa_vref is a global parameter (1 value per link)
+  std::array<double, 72> chan_effs;
+
+  // there is probably a better way to do the next line
+  std::array<std::array<double, 256>, 2> final_effs;
+  DecodeAndBuffer buffer{n_events}; // working in buffer, not in writer
+
+  // loop over runs, from toa_vref = 0 to = 255
+
+  for (int toa_vref{0}; toa_vref < 256; toa_vref++) {
+    pflib_log(info) << "testing toa_vref = " << toa_vref;
+    auto test_handle = roc.testParameters()
+      .add("REFERENCEVOLTAGE_0", "TOA_VREF", toa_vref)
+      .add("REFERENCEVOLTAGE_1", "TOA_VREF", toa_vref)
+      .apply();
+    usleep(10);
+    tgt->daq_run("PEDESTAL", buffer, n_events, 100);
+    pflib_log(trace) << "finished toa_vref = " << toa_vref << ", getting efficiencies";
+    auto efficiencies = get_toa_efficiencies(buffer.get_buffer());
+    chan_effs = efficiencies;
```

Comment:
It doesn't look like `chan_effs` is used anywhere? Maybe you can remove it now even though it may have been helpful during development.